torch_xla_py/xla_model.py

```diff
@@ -355,6 +360,29 @@ def xla_run_grad(xla_model, grad_outputs, devices=None):
   xla_model.backward(*grads_output_xla)
 
 
+def sync_optimizer(optimizer):
```

Comment:
This is hard-wired to dispatch the accumulated graph at the end of an optimizer step. Could we do something more general instead which would also work for a hook on forward? It should be possible to do this with no knowledge of the user code except for the fact that a training loop is complete. Again, not something for this PR, just a question for my own understanding.

Reply:
Sure, that is another way. Though, if the user wants to interact with the optimizer core object (calling other APIs, accessing its fields), his code needs to be changed in many places.

Reply:
Hooking the model is another place. Again, this means that is the user wants to interact with the model's other APIs/fields, he needs to change more.
With the optimizer_step(), only only line is changed.
But there are many ways we can go.
Bottom line, we need to "sync" parameters and gradients, in order to make a repeatable computation.

Reply:
Correction, model wrapping is not good (unless we pass optimizer - which is a chicken and egg problem), as we need to sync the optimizer's state as well (think, momentum).

Reply:
Yeah, these are good points. As long as there's no fundamental reason why this couldn't be more general, I'm not worried. We can always rework the Python APIs if we change our mind.